drivers/stepper/stepper_context.h

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2020 Teslabs Engineering S.L.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_DRIVERS_STEPPER_STEPPER_CONTEXT_H_
+#define ZEPHYR_DRIVERS_STEPPER_STEPPER_CONTEXT_H_
+
+#include <kernel.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct stepper_context {
+	struct k_sem lock;
+	struct k_sem sync;
+	int status;
+
+#ifdef CONFIG_STEPPER_ASYNC
+	struct k_poll_signal *signal;
+	bool asynchronous;
+#endif /* CONFIG_STEPPER_ASYNC */
+};
+
+#define STEPPER_CONTEXT_INIT_LOCK(_data, _ctx_name)                            \
+	._ctx_name.lock = Z_SEM_INITIALIZER(_data._ctx_name.lock, 0, 1)
+
+#define STEPPER_CONTEXT_INIT_SYNC(_data, _ctx_name)                            \
+	._ctx_name.sync = Z_SEM_INITIALIZER(_data._ctx_name.sync, 0, 1)
+
+static inline void stepper_context_lock(struct stepper_context *ctx,
+					bool asynchronous,
+					struct k_poll_signal *signal)
+{
+	k_sem_take(&ctx->lock, K_FOREVER);
+
+#ifdef CONFIG_STEPPER_ASYNC
+	ctx->asynchronous = asynchronous;
```

Comment:
I guess this could be OK, but it'd be messy if `asynchronous` was true while `signal == NULL`.